birdset/datamodule/embeddings_datamodule.py

```diff
@@ -0,0 +1,293 @@
+
+                # Convert the list of embeddings to a tensor
+                #embeddings = torch.stack(embeddings)
+
+                # Create a new Dataset with the embeddings
+                if len(embeddings) > 0:
+                    embeddings_dataset[split] =  Dataset.from_dict({key: [sample[key] for sample in embeddings] for key in embeddings[0]})
+                else:
+                    embeddings_dataset[split] = Dataset.from_dict({'audio':[], 'labels':[]})    
+
+        return embeddings_dataset
+        
+
+    def _get_embedding(self, audio):
+        # Get waveform and sampling rate
+        waveform = torch.tensor(audio['array'], dtype=torch.float32).to(self.device) # Get waveform audio and move to GPU
+        dataset_sampling_rate = audio['sampling_rate']
+        # Resample audio
+        audio = self._resample_audio(waveform, dataset_sampling_rate)
+        
+        # Zero-padding
+        audio = self._zero_pad(waveform)
+
+        # Check if audio is too long 
+        if waveform.shape[0] > self.max_length * self.sampling_rate:
+            if self.average:
+                return self._frame_and_average(waveform) 
+            else:
+                audio = audio[:self.max_length * self.sampling_rate]
+                return self.embedding_model.get_embeddings(audio.view(1, 1, -1))[0] 
+        else:
+            return self.embedding_model.get_embeddings(audio.view(1, 1, -1))[0] # To just use embeddings not logits
+
+    # Resample function
+    def _resample_audio(self, audio, orig_sr):
+        resampler = torchaudio.transforms.Resample(orig_freq=orig_sr, new_freq=self.sampling_rate)
+        return resampler(audio)
+
+    # Zero-padding function
+    def _zero_pad(self, audio):
+        desired_num_samples = self.max_length * self.sampling_rate 
+        current_num_samples = audio.shape[0]
+        padding = desired_num_samples - current_num_samples
+        if padding > 0:
+            #print('padding')
+            pad_left = padding // 2
+            pad_right = padding - pad_left
+            audio = torch.nn.functional.pad(audio, (pad_left, pad_right))
+        return audio
+
+    # Average multiple embeddings function
+    def _frame_and_average(self, audio):
+        # Frame the audio
+        frame_size = self.max_length * self.sampling_rate
+        hop_size = self.max_length * self.sampling_rate
+        frames = audio.unfold(0, frame_size, hop_size)
+        
+        # Generate embeddings for each frame
+        l = []
+        for frame in frames:
+            embedding = self.embedding_model.get_embeddings(frame.view(1, 1, -1)) 
+            l.append(embedding[0]) # To just use embeddings not logits
+        
+        embeddings = torch.stack(tuple(l))
+        
+        # Average the embeddings
+        averaged_embedding = embeddings.mean(dim=0)
+        
+        return averaged_embedding        
+    
+    def _save_dataset_to_disk(self, dataset: Dataset | DatasetDict):
```

Comment:
Whats the difference to `base_datamodule._save_dataset_to_disk`?

Reply:
It uses the custom created fingerprint (In init of the class) and in base_datamodule it creates/gets the fingerprint in the function

Reply:
@XgamerTV Ah okay, I guess it is cleaner to not use a new property if it can also be accessed by `self.embeddings_module.sampling_rate`. Please refactor this and also other chases where you use a new property as a "shortcut".

Reply:
It can't because only the EmbeddingModuleConfig saves it and self.embedding_model has the model itself not the config which is not saved, only the values needed from it are saved